clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleZonalServerGroupCachingAgent.java

```diff
@@ -0,0 +1,906 @@
+    } else {
+      String zone = Utils.getLocalName(manager.getZone());
+      serverGroup.setZone(zone);
+      serverGroup.setZones(ImmutableSet.of(zone));
+      serverGroup.setRegion(credentials.regionFromZone(zone));
+    }
+  }
+
+  private static List<String> getZones(DistributionPolicy distributionPolicy) {
+    if (distributionPolicy == null || distributionPolicy.getZones() == null) {
+      return ImmutableList.of();
+    }
+    return distributionPolicy.getZones().stream()
+        .map(z -> Utils.getLocalName(z.getZone()))
+        .collect(toImmutableList());
+  }
+
+  private static ImmutableMap<String, Integer> convertNamedPorts(InstanceGroupManager manager) {
+    if (manager.getNamedPorts() == null) {
+      return null;
+    }
+    return manager.getNamedPorts().stream()
+        .filter(namedPort -> namedPort.getName() != null)
+        .filter(namedPort -> namedPort.getPort() != null)
+        .collect(toImmutableMap(NamedPort::getName, NamedPort::getPort));
+  }
+
+  private static void setLaunchConfig(
+      GoogleServerGroup serverGroup,
+      InstanceGroupManager manager,
+      @Nullable InstanceTemplate instanceTemplate,
+      GoogleNamedAccountCredentials credentials,
+      ProviderCache providerCache) {
+
+    HashMap<String, Object> launchConfig = new HashMap<>();
+    launchConfig.put("createdTime", Utils.getTimeFromTimestamp(manager.getCreationTimestamp()));
+
+    if (instanceTemplate != null) {
+      List<AttachedDisk> disks = getDisks(instanceTemplate);
+      instanceTemplate.getProperties().setDisks(disks);
+      launchConfig.put("instanceTemplate", instanceTemplate);
+      launchConfig.put("launchConfigurationName", instanceTemplate.getName());
+      launchConfig.put("instanceType", instanceTemplate.getProperties().getMachineType());
+      launchConfig.put("minCpuPlatform", instanceTemplate.getProperties().getMinCpuPlatform());
+      setSourceImage(serverGroup, launchConfig, disks, credentials, providerCache);
+    }
+    serverGroup.setLaunchConfig(launchConfig);
+  }
+
+  private static List<AttachedDisk> getDisks(InstanceTemplate template) {
+
+    if (template.getProperties() == null || template.getProperties().getDisks() == null) {
+      return ImmutableList.of();
+    }
+    List<AttachedDisk> persistentDisks =
+        template.getProperties().getDisks().stream()
+            .filter(disk -> "PERSISTENT".equals(disk.getType()))
+            .collect(toImmutableList());
+
+    if (persistentDisks.isEmpty() || persistentDisks.get(0).getBoot()) {
+      return template.getProperties().getDisks();
+    }
+
+    List<AttachedDisk> sortedDisks = new ArrayList<>();
+    Optional<AttachedDisk> firstBootDisk =
+        persistentDisks.stream().filter(AttachedDisk::getBoot).findFirst();
+    firstBootDisk.ifPresent(sortedDisks::add);
+    template.getProperties().getDisks().stream()
+        .filter(disk -> !disk.getBoot())
+        .forEach(sortedDisks::add);
+    return sortedDisks;
```

Comment:
Could be immutable (particularly since the first `return` does return an immutable list).

Reply:
Absolutely, thanks.